File: src/main/java/com/pancisin/bookster/models/Invitation.java
package com.pancisin.bookster.models;

import java.util.Calendar;

import javax.persistence.Column;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.ManyToOne;
import javax.persistence.OneToOne;
import javax.persistence.Table;
import javax.persistence.UniqueConstraint;

import org.hibernate.validator.constraints.Email;

import com.fasterxml.jackson.annotation.JsonIgnore;
import com.fasterxml.jackson.annotation.JsonView;
import com.pancisin.bookster.models.views.Compact;

@Entity
@Table(name = "invitations", uniqueConstraints = { @UniqueConstraint(columnNames = { "email", "event_id" }) })
public class Invitation {

	@Id
	@JsonView(Compact.class)
	@GeneratedValue(strategy = GenerationType.AUTO)
	private Long id;

	@Email
	@Column
	private String email;

	@OneToOne(optional = true)
	private User user;

	@JsonView(Compact.class)
	@Column(name = "created", columnDefinition = "TIMESTAMP DEFAULT CURRENT_TIMESTAMP", insertable = false, updatable = false)
	private Calendar created;

	@JsonIgnore
	@ManyToOne(optional = true)
	private Event event;
	
	@JsonIgnore
	@ManyToOne(optional = true)
	private Conference conference;

	public Invitation() {
		
	}
	
	public Invitation(Event event, String email) {
		this.event = event;
		this.email = email;
	}
	
	public Invitation(Conference conference, String email) {
		this.conference = conference;
		this.email = email;
	}
	
	public String getEmail() {
		return email;
	}

	public void setEmail(String email) {
		this.email = email;
	}

	public User getUser() {
		return user;
	}

	public void setUser(User user) {
		this.user = user;
	}

	public Event getEvent() {
		return event;
	}

	public void setEvent(Event event) {
		this.event = event;
	}

	public Long getId() {
		return id;
	}

	public Calendar getCreated() {
		return created;
	}

	public Conference getConference() {
		return conference;
	}

	public void setConference(Conference conference) {
		this.conference = conference;
	}
}

File: src/main/kotlin/com/pancisin/bookster/model/Invitation.kt
package com.pancisin.bookster.model

import java.util.Calendar

import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.GeneratedValue
import javax.persistence.GenerationType
import javax.persistence.Id
import javax.persistence.JoinColumn
import javax.persistence.JoinTable
import javax.persistence.ManyToOne
import javax.persistence.OneToOne
import javax.persistence.Table

import org.hibernate.validator.constraints.Email

import com.fasterxml.jackson.annotation.JsonIgnore
import com.fasterxml.jackson.annotation.JsonView
import com.pancisin.bookster.models.views.Compact

@Entity
@Table(name = "invitations")
data class Invitation(

  @Id @JsonView(Compact::class) @GeneratedValue(strategy = GenerationType.AUTO)
  val id: Long? = null,

  @Email
  @Column
  var email: String? = null,

  @OneToOne(optional = true)
  var user: User? = null,

  @JsonView(Compact::class)
  @Column(name = "created", columnDefinition = "TIMESTAMP DEFAULT CURRENT_TIMESTAMP", insertable = false, updatable = false)
  val created: Calendar? = null,

  @JsonIgnore @ManyToOne
  @JoinTable(
    name = "events_invitations",
    joinColumns = arrayOf(JoinColumn(name = "invitation_id")),
    inverseJoinColumns = arrayOf(JoinColumn(name = "event_id"))
  )
  var event: Event? = null,

  @JsonIgnore @ManyToOne
  @JoinTable(
    name = "pages_invitations",
    joinColumns = arrayOf(JoinColumn(name = "invitation_id")),
    inverseJoinColumns = arrayOf(JoinColumn(name = "page_id"))
  )
  var page: Page? = null
)
